lib/srv/git/forward.go

```diff
@@ -0,0 +1,590 @@
+}
+
+// Dial returns the client connection of the pipe
+func (s *ForwardServer) Dial() (net.Conn, error) {
+	return s.clientConn, nil
+}
+
+// NewForwardServer creates a new in-memory SSH server that forwards git
+// commands to remote git-hosting services like "github.com".
+func NewForwardServer(cfg *ForwardServerConfig) (*ForwardServer, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	serverConn, clientConn, err := utils.DualPipeNetConn(cfg.SrcAddr, cfg.DstAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	logger := slog.With(teleport.ComponentKey, teleport.ComponentForwardingGit,
+		"src_addr", cfg.SrcAddr.String(),
+		"dst_addr", cfg.DstAddr.String(),
+	)
+	s := &ForwardServer{
+		StreamEmitter:    cfg.Emitter,
+		cfg:              cfg,
+		serverConn:       serverConn,
+		clientConn:       clientConn,
+		logger:           logger,
+		reply:            sshutils.NewReply(logger),
+		id:               uuid.NewString(),
+		waitExec:         make(chan error, 1),
+		verifyRemoteHost: verifyRemoteHost(cfg.TargetServer),
+		makeRemoteSigner: makeRemoteSigner,
+	}
+	// TODO(greedy52) extract common parts from srv.NewAuthHandlers like
+	// CreateIdentityContext and UserKeyAuth to a common package.
+	s.auth, err = srv.NewAuthHandlers(&srv.AuthHandlerConfig{
+		Server:       s,
+		Component:    teleport.ComponentForwardingGit,
+		Emitter:      s.cfg.Emitter,
+		AccessPoint:  cfg.AccessPoint,
+		TargetServer: cfg.TargetServer,
+		FIPS:         cfg.FIPS,
+		Clock:        cfg.Clock,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return s, nil
+
+}
+
+// Serve starts an SSH server that forwards git commands.
+func (s *ForwardServer) Serve() {
+	defer s.close()
+	s.logger.DebugContext(s.cfg.ParentContext, "Starting forwarding git")
+	defer s.logger.DebugContext(s.cfg.ParentContext, "Finished forwarding git")
+	server, err := sshutils.NewServer(
+		teleport.ComponentForwardingGit,
+		utils.NetAddr{}, /* empty addr, this is one time use so no use for listener*/
+		sshutils.NewChanHandlerFunc(s.onChannel),
+		sshutils.StaticHostSigners(s.cfg.HostCertificate),
+		sshutils.AuthMethods{
+			PublicKey: s.userKeyAuth,
+		},
+		sshutils.SetFIPS(s.cfg.FIPS),
+		sshutils.SetCiphers(s.cfg.Ciphers),
+		sshutils.SetKEXAlgorithms(s.cfg.KEXAlgorithms),
+		sshutils.SetMACAlgorithms(s.cfg.MACAlgorithms),
+		sshutils.SetClock(s.cfg.Clock),
+		sshutils.SetNewConnHandler(sshutils.NewConnHandlerFunc(s.onConnection)),
+	)
+	if err != nil {
+		s.logger.ErrorContext(s.cfg.ParentContext, "Failed to create git forward server", "error", err)
+		return
+	}
+	server.HandleConnection(s.serverConn)
+}
+
+func (s *ForwardServer) close() {
+	for _, closer := range []io.Closer{
+		s.serverConn,
+		s.clientConn,
+		s.cfg.TargetConn,
+	} {
+		if err := closer.Close(); err != nil && !utils.IsOKNetworkError(err) {
+			s.logger.WarnContext(s.cfg.ParentContext, "Failed to close", "error", err)
```

Comment:
Just copypaste this three times so the log message can include which connection had a problem?